### PR TITLE
Add VSM Setter TargetName

### LINKED ISSUE #4924: [F100] Add "Target" support to VisualStateManager
## Summary
This probably should be in this issue (https://github.com/xamarin/Xamarin.Forms/issues/4232), but I thought I would create a new one as this might be a smaller task.

I was porting my UWP app to Xamarin.Forms and I noticed that the VisualStateManager does not allow states to affect child controls. Take this pseudo xaml:

```xaml
<Grid>
    <VisualStateManager>
        <State Name="Big">
            <!-- New feature can change children -->
            <!-- UWP style -->
            <Setter Target="myButton.Color" Value="Red" />
            <!-- WPF style -->
            <Setter TargetName="myButton" Property="Color" Value="Red" />

            <!-- Current feature only changes the current element -->
           <Setter Target="BackgroundColor" Value="Red" />
        </State>
    </VisualStateManager>

    <Button x:Name="myButton" Color="Blue" />

</Grid>
```

This XAML allows the VSM to modify properties on the children, not just the current element.
https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.setter.target#Windows_UI_Xaml_Setter_Target
https://docs.microsoft.com/en-us/dotnet/api/system.windows.setter.targetname?view=netframework-4.7.2#System_Windows_Setter_TargetName

## API Changes

I am not sure if the UWP or the WPF route is better, but we might just be able to add a `TargetName` property to `Setter` and then instead of a `this.SetProperty(PropertyName, Value)` we would `this.FindElement(TargetName).SetProperty(PropertyName, Value)`.

## Intended Use Case

This will solve the issue of multiple VSM elements in a single page that responds to a single event, such as a state for a wide screen or a narrow screen, or for device orientation. One VSM can control the entire page.

The use case that I was working on was on a narrow screen, I would have 2 columns in my grid, but on a wide screen, I would have 4 and show a big header with more features.

### LINKED ISSUE #5622: CollectionView SelectedItem Color
### Description
When no BackgroundColor is set to ItemTemplate, the SelectedItem shows a Orange background color, when you set a BackgroundColor in ItemTemplate, then SelectedItem has the same color than the rest of the items and you could not Identify the SelectedItem.
It would be appreciated also being able to set the BackgroundColor / BorderColor of selectedItem instead of just a orange color.
### Steps to Reproduce

Having this XAML
```
<ContentPage>
        <Grid>
            <Grid.ColumnDefinitions>
                <ColumnDefinition />
                <ColumnDefinition Width="auto" />
            </Grid.ColumnDefinitions>
            <CollectionView Grid.Column="0" 
                            SelectionMode="Single"
                            SelectedItem="{Binding SelectedItem, Mode=TwoWay}"
                            ItemsSource="{Binding ObservableCollectionTest}">
                <CollectionView.ItemsLayout>
                    <GridItemsLayout Orientation="Vertical" />
                </CollectionView.ItemsLayout>
                <CollectionView.ItemTemplate>
                    <DataTemplate>
                        <Grid BackgroundColor="Wheat">
                            <Label Text="{Binding}"></Label>
                        </Grid>
                    </DataTemplate>
                </CollectionView.ItemTemplate>
                
            </CollectionView>
            <Label Grid.Column="1"
                   Text="{Binding SelectedItem}" />
        </Grid>
    </ContentPage>
```
With this data in the Model
```
public static string SelectedItemPropertyName() { return nameof(SelectedItem); }
        private string _SelectedItem;
        public string SelectedItem
        {
            get
            {
                return _SelectedItem;
            }
            set
            {
                _SelectedItem = value;
                RaisePropertyChanged();
            }
        }
        public static string ObservableCollectionTestPropertyName() { return nameof(ObservableCollectionTest); }
        private ObservableCollection<string> _ObservableCollectionTest = new ObservableCollection<string>(new List<string>() { "Test1", "test2", "test3" } );
        public ObservableCollection<string> ObservableCollectionTest
        {
            get
            {
                return _ObservableCollectionTest;
            }
            set
            {
                _ObservableCollectionTest = value;
                RaisePropertyChanged();
            }
        }
SelectedItem = ObservableCollectionTest[1];
```
### Expected Behavior
SelectedItem having a different background color or at least a border color
### Actual Behavior
It looks like the rest of the items
### Basic Information

- Version with issue: 4.0.0.232914-pre6

- IDE: Visual Studio Community 2017 version 15.9.9
- Platform Target Frameworks: 
  - Android: 8.1

### PR DESCRIPTION
**Pull request by jfversluis**
### Description of Change ###

Adds the ability to target a child object by name for a Setter for visual state management or styles.

### Issues Resolved ### 

- fixes #4924 
- closes #5034
- closes #5622

### API Changes ###

Added:
 - `public string TargetName { get; set; }`

### Platforms Affected ### 

- Core/XAML (all platforms)

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###

I've added a core gallery page `VisualStateSetterTarget.xaml` which has a button to toggle the VSM state of the `StackLayout`. When it's pressed, the `ALabel` is also triggered because of it's reference through `TargetName`.

### PR Checklist ###

- [x] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Xamarin.Forms.Core/Setter.cs [hunk 1]
```diff
@@ -47,59 +48,73 @@ object IValueProvider.ProvideValue(IServiceProvider serviceProvider)
 			return this;
 		}
 
-		internal void Apply(BindableObject target, bool fromStyle = false)
+		internal void Apply(BindableObject scope, bool fromStyle = false)
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Setter.cs]: the name was correct

File: Xamarin.Forms.Core/Setter.cs [hunk 2]
```diff
@@ -47,59 +48,73 @@ object IValueProvider.ProvideValue(IServiceProvider serviceProvider)
...
 		{
-			if (target == null)
-				throw new ArgumentNullException(nameof(target));
+			var targetObject = FindTargetObject(scope, TargetName);
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Setter.cs]: I think the FindTargetObject doesn't bring enough value by itself
```csharp
if (target == null) throw...;
if (!string.IsNullOrEmpty(TargetName) && target is Element element)
    target == element.FindByName(TargetName) as BindableObject ?? throw ...;
```

Comment (jfversluis) [reply Xamarin.Forms.Core/Setter.cs]: I think I implemented the way you intended, if I missed anything please let me know :)

**Comment by samhouts** [comment]
please target 4.4.0

**Comment by jfversluis** [comment]
🤦‍♂ done!

**Comment by mattleibow** [comment]
Does this support property paths? For example:

```
<Setter TargetName="ItemGrid" Property="ColumnDefinitions[2].Width" Value="0" />
```

I have an app that actually does this: https://gist.github.com/mattleibow/2f2f33c353be0b33e31bf9007095b105#file-mainpage-xaml-L35-L41

**Comment by jfversluis** [comment]
> Does this support property paths? For example:
> 
> ```
> <Setter TargetName="ItemGrid" Property="ColumnDefinitions[2].Width" Value="0" />
> ```
> 
> I have an app that actually does this: https://gist.github.com/mattleibow/2f2f33c353be0b33e31bf9007095b105#file-mainpage-xaml-L35-L41

That is pretty awesome :) 

Just tried it with this PR and that doesn't work. Would it make sense to get this in first as a start and then build from there?

**Comment by mattleibow** [comment]
That seems like something g that we can do still. I can always name things a bit more for now. Good work so far.

Gonna be nice to write less VSM per element :)

**Pull request merged**